### ISSUE TITLE
[ls] Convenient support for custom `InternalStatefulIncrementalBuilder` bindings

### ISSUE THREAD

**Issue by kittaakos** [eclipse/xtext-core#798]
Currently, I need to do something like this if I would like to have a custom binding for my LS. 
```java
class MyServerModule extends ServerModule {

  override protected configure() {
    super.configure();
    binder.bindInternalStatefulIncrementalBuilder;
  }

  protected def bindInternalStatefulIncrementalBuilder(Binder binder) {
    val fromClass = IncrementalBuilder.classLoader.loadClass(
      'org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder') as Class<Object>;
    (binder.bind(fromClass) as AnnotatedBindingBuilder<Object>).to(MyInternalStatefulIncrementalBuilder);
  }

}
```

The `InternalStatefulIncrementalBuilder` class is a protected static class of the `IncrementalBuilder`. Please make it `public`, or I can provide a PR too.

**Comment by svenefftinge** [comment]
I don't remember. Sounds good to me to expose it.

**Comment by kittaakos** [comment]
Great, thanks for checking. There is another thing I would like to modify within this PR:

https://github.com/eclipse/xtext-core/blob/909bf238d109ddb8ea1faceb1d18fe6c4500aeac/org.eclipse.xtext/src/org/eclipse/xtext/build/IncrementalBuilder.xtend#L53-L54

These two fields have their generated `protected` setters, but they're not used because of the assignments

https://github.com/eclipse/xtext-core/blob/909bf238d109ddb8ea1faceb1d18fe6c4500aeac/org.eclipse.xtext/src/org/eclipse/xtext/build/IncrementalBuilder.xtend#L222-L223

happen in a static subclass, so that they will be direct field assignments instead of calls through the setters. Long story short, overriding classes cannot access the `buildContext` so I either make both the `buildContext` and the `request` protected fields or change the assignments to

```java
  builder.setContext(context)
  builder.setRequest(request)
```

**Comment by cdietrich** [comment]
Fixed in 2.15
